reference/factorization/par_ic_kernels.cpp

```diff
@@ -0,0 +1,140 @@
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include "core/factorization/par_ic_kernels.hpp"
+
+
+#include <ginkgo/core/base/math.hpp>
+#include <ginkgo/core/matrix/coo.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+
+
+#include "core/base/utils.hpp"
+
+
+namespace gko {
+namespace kernels {
+namespace reference {
+/**
+ * @brief The parallel ic factorization namespace.
+ *
+ * @ingroup factor
+ */
+namespace par_ic_factorization {
+
+
+template <typename ValueType, typename IndexType>
+void init_factor(std::shared_ptr<const DefaultExecutor> exec,
+                 matrix::Csr<ValueType, IndexType> *l)
+{
+    auto num_rows = l->get_size()[0];
+    auto l_row_ptrs = l->get_const_row_ptrs();
+    auto l_col_idxs = l->get_const_col_idxs();
+    auto l_vals = l->get_values();
+
+    for (size_type row = 0; row < num_rows; ++row) {
+        for (size_type l_nz = l_row_ptrs[row]; l_nz < l_row_ptrs[row + 1];
+             ++l_nz) {
+            if (l_col_idxs[l_nz] == row) {
+                auto diag = sqrt(l_vals[l_nz]);
+                if (is_finite(diag)) {
+                    l_vals[l_nz] = diag;
+                } else {
+                    l_vals[l_nz] = one<ValueType>();
+                }
+            }
+        }
+    }
+}
+
+GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
+    GKO_DECLARE_PAR_IC_INIT_FACTOR_KERNEL);
+
+
+template <typename ValueType, typename IndexType>
+void compute_factor(std::shared_ptr<const DefaultExecutor> exec,
+                    size_type /* num_iterations */,
+                    const matrix::Coo<ValueType, IndexType> *a_lower,
+                    matrix::Csr<ValueType, IndexType> *l)
+{
+    auto num_rows = a_lower->get_size()[0];
+    auto l_row_ptrs = l->get_const_row_ptrs();
+    auto l_col_idxs = l->get_const_col_idxs();
+    auto l_vals = l->get_values();
+    auto a_vals = a_lower->get_const_values();
+
+    for (size_type row = 0; row < num_rows; ++row) {
+        for (size_type l_nz = l_row_ptrs[row]; l_nz < l_row_ptrs[row + 1];
+             ++l_nz) {
+            auto col = l_col_idxs[l_nz];
+            auto a_val = a_vals[l_nz];
+            // accumulate l(row,:) * l(col,:) without the last entry l(col, col)
```

Comment:
Is there any reason we do the last entry computation and then remove it in ParILU?
Or, is it just different implementation style?

Reply:
Yes, I added the `l_col < last_entry` condition to remove the last term